README.md

```diff
@@ -19,33 +26,239 @@ yarn build
+| ------ | ------ | --------------------- | -------- | -------------------------------- |
+| target | string | "^[a-zA-Z0-9_-]{43}$" | true     | Address to retrieve balance for. |
+
+```typescript
+const { originalTxId } = await contract.viewState({
+  function: 'balance',
+  target: '000000000000000000000000000000000000000000', // the address to view the balance of
+});
+```
+
+### `evolve`
+
+Allows the contract to evolve by setting a new contract source.
+
+| Name  | Type   | Pattern               | Required | Description                       |
+| ----- | ------ | --------------------- | -------- | --------------------------------- |
+| value | string | "^[a-zA-Z0-9_-]{43}$" | true     | New source code for the contract. |
+
+```typescript
+const { originalTxId } = await contract.writeInteraction({
+  function: 'evolve',
+  value: '000000000000000000000000000000000000000000', // the source code transaction to use
+});
+```
+
+# Development
+
+### Using AJV for Input Validation in Contract Methods
+
+We employ AJV (Another JSON Schema Validator) to ensure the integrity and correctness of inputs for our smart contract methods. AJV is a fast JSON schema validator that allows us to define the expected format of inputs and validate these inputs against the defined schemas.
+#### Overview of AJV Integration
+
+We use AJV to compile JSON schemas into validation functions, which are then used to validate inputs for our contract methods. The schemas define the expected structure, type, and constraints of the data. Our build file (`build.js`) includes the setup and compilation of these validation functions.
+
+#### Schemas
+
+We have defined multiple schemas for different contract methods, such as `balanceSchema`, `setControllerSchema`, `setNameSchema`, etc. These schemas are located in the `./schemas` directory. Each schema precisely defines the expected format of inputs for a corresponding contract method.
+
+#### Compilation of Validation Functions
+
+In the build process, we create an instance of AJV and add our schemas to it. We then use `standaloneCode` from `ajv/dist/standalone` to compile these schemas into standalone validation functions. This compilation process generates a module (`validations.js`) in the `src` directory, which contains all the validation functions.
+
+#### Usage in Contract Methods
+
+In our contract methods, we import the relevant validation function from `validations.js` and use it to validate the input data. For example, if we have a contract method `setRecord`, we use the `validateSetRecord` function to validate its inputs.
+
+```javascript
+const { validateSetRecord } = require('./validations');
+
+function setRecord(input) {
+  if (!validateSetRecord(input)) {
+    throw new Error('Invalid input');
+  }
+}
+```
+
+# Additional Resources
+
+### Documentation
+
+- [Arweave Name System Documentation]
+- [Permaweb Cookbook]
+- [AJV]
```

Comment:
Add links to portal app and arns service repo

Reply:
added